### ISSUE TITLE
Add actual LiveReload support

### ISSUE THREAD

**Issue by fbricon** [quarkusio/quarkus#1418]
I already mentioned it in zulip some time ago, but apparently forgot to open an issue so here goes: in spring boot, if the spring dev tools dependency is available, a [livereload](http://livereload.com/) server instance is started. changes to files trigger a livereload event that makes all clients (i.e browsers) to refresh automagically.

Now, in quarkus, if you could listen to a "classloader was reloaded" event, you could tell a livereload server to propagate the reload event to all clients, when java files are changed, so you wouldn't need to manually refresh the browser (it's way much cooler to perform a demo when multiple screens are connected to the same server). Listening to non-class changes could also trigger the reload (i.e html, js, css...)

**Comment by danielpetisme** [comment]
Same idea #1410

**Comment by gsmet** [comment]
So, I'm not sure it's a good thing for Java files (some people were concerned about reloading partial changes that wouldn't compile and generate a lot of noise in the log...).

But I think we should do it for static resources as you could at least do the CSS changes in live.

**Comment by fbricon** [comment]
That's why I mentioned the "classloader was reloaded" event

**Comment by Sanne** [comment]
I love it conceptually, but we'll need to address another design aspect too:

the "reload" operation today is waiting for some kind of trigger. We don't want to "reload" the whole app just because a single file changed, especially as a change might be in progress - and not ready to be compiled yet.

So today the browser refresh is actually the event that triggers re-compile, process it all, reload the app ..

If it's not the browser, what should it be? A git commit hook maybe? a button in your IDE?
I'm thinking it could be the "touching" of a specific file, so that people can hook in anything they like.

Another alternative would be to have Quarkus watch the file changes, and simply keep trying - to only reload when there are no compilation errors. But would people like this? You'd have no feedback at all on mistakes.

N.B. I usually don't use the browser so I'm not at all defending the current design as I'd love to use it without necessarily having a browser open ;)

@stuartwdouglas is the master of puppets here.

**Comment by fbricon** [comment]
One would have to check how springboot handles it, but I imagine one solution could be: 
- detect file changed
- wait for X ms for more changes
- changes seem to be over => trigger reload.

It might not be perfect, but will prolly handle more than 80% of the use cases. And you could still add a flag to disable the behavior

**Comment by geoand** [comment]
Honestly I find the Quarkus approach to live reload much more intuitive for developing APIs that what Spring Boot does. It just seems to fit the API developer workflow like a glove!

Personally the only thing I would like would be to have reload work for event based applications as well.

Reloading the web assets can certainly be done using the current approach as Stuart has demonstrated in an old PR.

**Comment by stuartwdouglas** [comment]
We could certainly offer it as an option and see how it actually works in practice, however I would also want to keep the current behaviour. 

The issue is picking a value of X that makes sense for the poll time and the 'wait X ms logic'. If you make this a small value then the FS is being hit constantly, which can impact IDE performance. If you make this a large value then when you refresh there is a good chance you will see out of date content.

I also think that livereload makes way more sense than CSS + HTML than it does for java. When modifying these assets the changes are likely visual, so when the browser updates you can see the effects immediately.  With Java backend code you often need to hit a button or perform some other action to see the result. Depending on how stateful your app is auto reloading could actually be a drain on productivity.

**Comment by vorburger** [comment]
> Personally the only thing I would like would be to have reload work for event based applications as well.

#641

**Comment by geoand** [comment]
> > Personally the only thing I would like would be to have reload work for event based applications as well.
> 
> #641

True, but I also had Kafka messages in mind :)

**Comment by emmanuelbernard** [comment]
A couple of ideas:
1. adding other input as the trigger for live reload (like websocket) is likely our main way forward
2. as part of the discussion around front end development support (with resources hosted by quarkus), then indeed the hook to livereload makes sense but indeed we need to filter which resource trigger that vs which does not (like java files)

It's a bit unclear to how to best address 2.

**Comment by dector** [comment]
Is it possible to re-use some existing Quarkus file watching mechanism in a reactive way to detect if source files or resources were changed? Or it's not yet possible and we need to implement our own file-watcher for now?

**Comment by Sanne** [comment]
@dector probably best to ask on zulip [https://quarkusio.zulipchat.com/#narrow/stream/187038-dev].

I think you could use `WatchServiceFileSystemWatcher` but it's not meant as public API - @stuartwdouglas might know about something better.

I'd also question why you need it? In general Quarkus just rebuilds it all, as we can do it quickly.

**Comment by dector** [comment]
@Sanne thanks for your reply!

Currently I'm evaluating quarkus as my next main BE development stack for small/medium project.
I'm quite impressed with it's functionality and simplicity ("hardcore enterprise" nature is still very obvious here and there but in acceptable amounts for now).

At the current point one thing I'm missing - proper live reload. Classes hot reload works awesome BTW but development UX is much more pleasant in my case if I don't need to refresh browser manually. So I've implemented live reload using web socket on BE + Violentmonkey script in browser. But I'm using additional dependency for recursive file-watching. Would be nice to be able to remove this dependency if quarkus has some native functionaly for file-watching.

[Demo](https://i.imgur.com/T7exbEu.mp4)

**Comment by Sanne** [comment]
I must say it looks awesome.

There's a conceptual problem with such an approach though: often one single consistent "patch" requires changes in multiple files to compile successfully - if you trigger a live reload just after a single file has changed, you might not have all the changes yet - this would most likely result in a sequence of reload events, some of the intermediary ones failing as the project won't necessarily compile.

Please see https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/.22real.22.20live.20reload

**Comment by geoand** [comment]
As @Sanne mentioned, the [Zulip Thread](https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/.22real.22.20live.20reload) contains a lot of details about why we would not want to change the live reload feature, so I'll close this broad scoped issue as `won't fix`.
We can of course consider improvements (@Ladicek has something in mind in the Zulip thread) but it would be best to have topical issues for those.

**Comment by Sanne** [comment]
Actually we'll re-open as @fbricon also suggested the browser could trigger live-reload on web resources.

That's still an applicable improvement which sound very useful - I'll change the title to reflect.

**Comment by dector** [comment]
> There's a conceptual problem with such an approach though: often one single consistent "patch" requires changes in multiple files to compile successfully

@Sanne that's true. My solution was intended to create short feedback loop for changing templates (_BTW Qute seems like the best template engine I've ever worked!_) - I'm using Tailwind and such primitive live-reload helps me to experiment/iterate on design very quick. It was quite easy to implement and is very helpful in my case. Most probably I'll need to extract file watcher into separate utility because now live-reload doesn't work if website can't be compiled.

@Sanne @geoand thanks, I'll check the thread

**Comment by blazmrak** [comment]
I use gulp + browser-sync for this. I know that this isn't a Quarkus solution, but it works quite nice for me, because I have to have tailwind installed anyways. It basically just starts a proxy to your server and listen to file changes. Each time the change is detected, the browser will reload the page after some delay.

1. install the dependencies:
```bash
npm i -D gulp gulp-cli browser-sync
```

2. confirm that gulp cli works
```bash
npx gulp --version
```

3. Create gulp script (edit the glob pattern in `gulp.watch` to match your project) 
```javascript
// gulpfile.js

const gulp = require('gulp');
const browserSync = require('browser-sync').create();

exports.build = gulp.task('sync', function() {
    browserSync.init({
        proxy: {
            target: 'http://127.0.0.1:8080', // this is the server, localhost didn't work for me for some reason
        },
        ui: { port: 3003 }, // UI, can be any port
        reloadDelay: 200 // Important, otherwise syncing will not work
    });
    gulp.watch([
        '../public/**/*.js',
        '../public/**/*.html',
        '../public/**/*.css',
        '../src/main/**',
    ]).on("change", browserSync.reload);
});
```

4.  Run `npx gulp sync`.

**Comment by vorburger** [comment]
Regarding:

> Most probably I'll need to extract file watcher into separate utilit

and:

> I think you could use `WatchServiceFileSystemWatcher` but it's not meant as public API - @stuartwdouglas might know about something better.

Just FYI https://github.com/vorburger/ch.vorburger.exec in case that could ever be of any interest to anyone in this context?

**Comment by stuartwdouglas** [comment]
WatchServiceFileSystemWatcher was copied directly from XNIO, it should be fine to use, but it certainly not something we ever claimed was a stable API. It's only one class though so if anything ever did change you could just copy it.

**Comment by simasch** [comment]
I did some experiments with JSF (Primefaces) and Quarkus and the browser reload is really missing when working on the xthml and CSS files.

**Comment by maxandersen** [comment]
context: https://twitter.com/maxandersen/status/1767655187819225330

@ia3andy didn't you do something in the area of handling updates to js/css/html ? feels like we at least should be able to offer trigger browse livereload for some resources...possibly adopt https://github.com/38leinaD/quarkus-livereload-extension to more recent version ?

**Comment by mstahv** [comment]
Remingind you @maxandersen that we discussed this a while ago when we were finishing the Vaadin integration. The dev mode in Vaadin does exactly that, automatically reloads the browser that web developers usually have on a second screen on "hot reload"/save/compile (in Spring Boot with Dev Tools or with basic servlet/CDI apps using JRebel/HotSwapAgent). Excellent for productivity as one don't need to leave the IDE to see changes in the UI (potentially on multiple screens/windows/devices). This is proably why @simasch has this level as gold standard, although he right now seems to be on the "dark side" with JSF 🤓

IIRC there was nothing in place where the Vaadin devmode could have connected as all checks for new changes are done lazily when a request starts in Quarkus dev mode. But if this has changed/will change, there we have a project with requirements 🙋‍♂️

**Comment by simasch** [comment]
@mstahv it is not the dark side. It makes me reminisce 😅

But you are right I use Vaadin with JRebel and that sets the bar high.

**Comment by maxandersen** [comment]
@mstahv yes, I fully get the value on it. having participated in implementing reload and livereload support for JBoss in Eclipse eons ago I get the value - just that quarkus default focus (good UX for backend dev where we only spend cycles on actual request) the "refresh browser pro-actively" have not been done. @simasch tweets made me realize we could do it without changing the default ux by doing something akin to how browser-sync does it. might not be 100% incremental as some state might be lost - but at least the browser does refresh and reloads on updates.

**Comment by GuillaumeTaffin** [comment]
If I may add to this thread...

Today the [HTMX](https://htmx.org/) lib is getting a lot in popularity, which means that server side rendering is not only something that "javascript web" framework do, but golang/rust/python/java... devs can also benefit from it!

Having the dev mode that refreshes the whole app but requiring to actually refreshing manually the browser to se the effect is **very** annoying as - in that context - I do it many times per minute

**Comment by blazmrak** [comment]
@GuillaumeTaffin I think Renarde and Web Bundler already implement this :) From what I vaguely remember, Renarde uses web bundler and web bundler is the one which actually implements live reload on file change :)

**Comment by ia3andy** [comment]
Yes: https://docs.quarkiverse.io/quarkus-web-bundler/dev/advanced-guides.html#web-live-coding

**Comment by ia3andy** [comment]
Also note that the web-bundler is a perfect match for htmx :)

**Comment by ia3andy** [comment]
I need to refresh it with new versions https://github.com/ia3andy/htmx-todo but this should give you some ideas.

**Comment by GuillaumeTaffin** [comment]
@blazmrak @ia3andy thank you for sharing !

**Comment by jnbdz** [comment]
I've built a prototype of what @maxandersen described in 2024 — browser-sync-style, opt-in, without touching the default dev-mode UX — and I'd like a maintainer read on whether core wants it before opening a PR, since this issue was once closed as won't-fix and @ia3andy's answer so far has been the web-bundler. Branch with the working change, tests and docs: https://github.com/SiteNetSoft/quarkus/tree/http-browser-live-reload (one commit, +458/−1, all in `quarkus-vertx-http`).

What it is, scoped to what @gsmet already agreed with in the very first comment ("we should do it for static resources"):

- Opt-in: `quarkus.http.browser-live-reload=true`, dev mode only. Off by default, nothing changes for anyone else, and Java-change restarts do **not** reload the browser — only changes that don't require a restart are pushed, so the compile-error noise concern from the original discussion doesn't apply.
- A file system watcher (the existing `WatchServiceFileSystemWatcher`, inotify-based — not polling, addressing @stuartwdouglas's 2019 concern) on the `META-INF/resources` source directories. It does nothing except trigger the regular `doScan`; the changed files then reach the browsers through the existing `consumeNoRestartChanges` hot-replacement notification — the exact path the web-bundler uses, minus the bundling. Because it rides that channel, other no-restart changes (e.g. Qute templates) also produce a reload, which is what the htmx users above are asking for.
- An SSE stream at `/q/browser-live-reload` and a ~40-line script at `/q/browser-live-reload.js`: a changed style sheet is swapped in place, anything else is `location.reload()`. No injection into responses — the page includes the script tag, which is inert in a packaged application since neither route exists outside dev mode.

Relationship to the web-bundler: this is deliberately the thin slice for applications that just serve `META-INF/resources` (or render server-side) and don't want bundling. The web-bundler keeps doing everything else it does; the SSE endpoint and script here are core-only and don't clash with its own (`/web-bundler/live-reload`).

If this shape is acceptable I'll open the PR; if the position is still "web-bundler is the answer", that's a fine outcome too and this issue could then be closed with that as the documented answer.
